### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by rlustin**
It adds the basis for authentication with a token in HTTP headers:
- a session controller to login (and get a session token) and logout;
- a middleware to protect the API behind authentication;
- a user controller to manage user (just create for now).

@Signez, could you please review?

### CODE REVIEW

File: src/authentication/middleware.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            response: Response::with(status::Unauthorized),
+        })
+    }
+}
+
+impl BeforeMiddleware for AuthenticationMiddleware {
+    fn before(&self, request: &mut Request) -> IronResult<()> {
+        use iron::headers::*;
+
+        // FIXME: find a more elegant way to match login path
```
Comment (Signez) [inline src/authentication/middleware.rs]: What about a `match` statement with `request.url.path().last()` and `Ok("login")`? Does it work?

Comment (rlustin) [inline src/authentication/middleware.rs]: It works with `request.url.path().last() == Some(&"login")` test.

**Comment by rlustin** [comment]
IMO, the real [Travis error](https://travis-ci.org/rlustin/lugh/builds/182829279#L319) we need to focus on is the database lock:
```
Error saving new user: DatabaseError(__Unknown, "database is locked")
```

I think that all other errors encountered in Travis are a consequence of this lock, depending on the order of execution of the tests.

For debugging, I printed in Travis the [options](https://travis-ci.org/rlustin/lugh/builds/182830108#L273-L282) used to compile sqlite. As we don’t have the `HAVE_USLEEP` flag, it may lead to wait seconds instead of milliseconds according to the [documentation](https://www.sqlite.org/compile.html#usleep).

I’ve [increased](https://github.com/rlustin/lugh/pull/18/commits/fa59bf9478d108efa5f2edaf23c5b3a01c3bba65) the busy timeout parameter up to 30 seconds, and I always have this lock.

What do you think we can do?

**Comment by rlustin** [comment]
However tests run perfectly on my machine…

**Comment by Signez** [comment]
I just tried to enforce `trusty` (_Ubuntu 14.04 LTS_, that came out in 2014) instead of the Travis-default `precise` environment.

And tests passes [all green](https://travis-ci.org/rlustin/lugh/builds/182854966). ✅ 

I invite you to add [this little change](https://github.com/rlustin/lugh/commit/2c4292dd6258731a6e6f812ee6ee0de3a36ea2d4), revert back some of those hacks, and enjoy the sanity comeback that will follow. 😉

**Comment by rlustin** [comment]
:bananadance:

That’s a very satisfying way of fixing :) Thank you @Signez!

I’ve updated the branch, ready for review!

**Pull request merged**